Converts arbitrary energy units to the joule. Fixes conservation of energy issues relating to charging cells.

Pickle-Coding opened the pull request:
## About The Pull Request
Removes all arbitrary energy and power units in the codebase. Everything is replaced with the joule and watt, with 1 = 1 joule, or 1 watt if you are going to multiply by time. This is a visible change, where all arbitrary energy units you see in the game will get proper prefixed units of energy.

With power cells being converted to the joule, charging one joule of a power cell will require one joule of energy.

The grid will now store energy, instead of power. When an energy usage is described as using the watt, a power to energy conversion based on the relevant subsystem's timing (usually multiplying by seconds_per_tick or applying power_to_energy()) is needed before adding or removing from the grid. Power usages that are described as the watt is really anything you would scale by time before applying the load. If it's described as a joule, no time conversion is needed. Players will still read the grid as power, having no visible change.

Machines that dynamically use power with the use_power() proc will directly drain from the grid (and apc cell if there isn't enough) instead of just tallying it up on the dynamic power usages for the area. This should be more robust at conserving energy as the surplus is updated on the go, preventing charging cells from nothing.

APCs no longer consume power for the dynamic power usage channels. APCs will consume power for static power usages. Because static power usages are added up without checking surplus, static power consumption will be applied before any machine processes. This will give a more truthful surplus for dynamic power consumers.

APCs will display how much power it is using for charging the cell. APC cell charging applies power in its own channel, which gets added up to the total. This will prevent invisible power usage you see when looking at the power monitoring console.

After testing in MetaStation, I found roundstart power consumption to be around 406kW after all APCs get fully charged. During the roundstart APC charge rush, the power consumption can get as high as over 2MW (up to 25kW per roundstart APC charging) as long as there's that much available.

Because of the absurd potential power consumption of charging APCs near roundstart, I have changed how APCs decide to charge. APCs will now charge only after all other machines have processed in the machines processing subsystem. This will make sure APC charging won't disrupt machines taking from the grid, and should stop APCs getting their power drained due to others demanding too much power while charging. I have removed the delays for APC charging too, so they start charging immediately whenever there's excess power. It also stops them turning red when a small amount of cell gets drained (airlocks opening and shit during APC charge rush), as they immediately become fully charged (unless too much energy got drained somehow) before changing icon.

Engineering SMES now start at 100% charge instead of 75%. I noticed cells were draining earlier than usual after these changes, so I am making them start maxed to try and combat that.

These changes will fix all conservation of energy issues relating to charging powercells.
## Why It's Good For The Game
Closes #73438
Closes #75789
Closes #80634
Closes #82031

Makes it much easier to interface with the power system in the codebase. It's more intuitive. Removes a bunch of conservation of energy issues, making energy and power much more meaningful. It will help the simulation remain immersive as players won't encounter energy duplication so easily. Arbitrary energy units getting replaced with the joule will also tell people more meaningful information when reading it. APC charging will feel more snappy.
## Changelog
:cl:
fix: Fixes conservation of energy issues relating to charging powercells.
qol: APCs will display how much power they are using to charge their cell. This is accounted for in the power monitoring console.
qol: All arbitrary power cell energy units you see are replaced with prefixed joules.
balance: As a consequence of the conservation of energy issues getting fixed, the power consumption for charging cells is now very significant.
balance: APCs only use surplus power from the grid after every machine processes when charging, preventing APCs from causing others to discharge while charging.
balance: Engineering SMES start at max charge to combat the increased energy loss due to conservation of energy fixes.
/:cl:

Pickle-Coding committed:
[WIP]Converts all the "energy unit" units to joule. Fixes conservation of power issues.

Converts all those arbitrary energy units across the codebase to the joule.

The WATT and JOULE defines are removed. I have no idea what their purpose were, but we don't need them.

Fixes as many conservation of energy issues as found, except for heat conversion and magic tree shit. If energy is given, it should be taken in equal, if not more.

Replaces some things with defines. Cleans some crap up. It probably would've been easier to be detailed if I wrote the commit while writing the changes. Apparantly the standard cell charge define was in the projectiles defines file lol.

There's a lot more things to convert. I haven't even tested this.

Kapu1178 commented:
add a KILOJOULES define that is just `* 1000`, it'll make numbers alot prettier

Pickle-Coding commented:
> add a KILOJOULES define that is just `* 1000`, it'll make numbers alot prettier

Seems kinda specific considering the range of values seen can be vastly higher than that.

Pickle-Coding committed:
Skill issue.

Pickle-Coding commented on code/game/machinery/defibrillator_mount.dm:
```diff
@@ -196,12 +197,12 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/machinery/defibrillator_mount, 28)
-		use_power(active_power_usage * seconds_per_tick)
-		C.give(40 * seconds_per_tick)
+	if(cell.charge < cell.maxcharge)
+		var/power_used = cell.give(active_power_usage * seconds_per_tick)
+		use_power(power_used)
```
This doesn't catch all cases. I'll do this properly later.

SyncIt21 commented on code/__DEFINES/power.dm:
```diff
@@ -11,9 +11,12 @@
+//#define JOULES * 0.002
+
+///The amount of energy, in joules, a standard powercell has.
+#define STANDARD_CELL_CHARGE 1e6
```
1e6 is confusing. You should do 1000 JOULES. Where JOULES is a defi e can be typed as 1e3 so people who aren't good at match can understand better

Pickle-Coding replied:
Why would JOULES be equal to 1,000? 1 = 1 joule.

Pickle-Coding replied:
I realised I forgot to remove commented out code. You can ignore that.

SyncIt21 replied:
Even though 1 joule is = 1 watt then create a define like

`# define JOULE * 1` and you can go all the way to giga joule.

Trush me, a little bit of semantics can improve readibility a long way

Pickle-Coding replied:
The define's comment specifies it's in joules.

SyncIt21 replied:
Comments are insufficient and error prone to changes.

We have defines like DESICOND whose value is also 1. Same is needed for joules

Pickle-Coding replied:
I'll define the JOULE, JOULES, WATT and WATTS simply to state they're the default energy and power units. Seems like people assumed they are considering at least 3 of the conservation of energy issues were related to people assuming powercells used joules.

SyncIt21 commented on code/game/objects/items/inspector.dm:
```diff
@@ -29,7 +29,7 @@
 	var/power_per_print = INSPECTOR_POWER_USAGE_NORMAL
 	///Power used to say an error message
-	var/power_to_speak = 1
+	var/power_to_speak = 1e3
```
1 JOULES, like my above comment

SyncIt21 commented on code/modules/modular_computers/computers/machinery/modular_computer.dm:
```diff
@@ -140,7 +140,7 @@
 	if(isnull(cell) || cell.percent() >= 100)
 		return
-	var/power_to_draw = idle_power_usage * seconds_per_tick * 0.5
+	var/power_to_draw = idle_power_usage * seconds_per_tick * 0.5 // Why halve it?
```
Because seconds_per_tick is by default 2 so halving it would become 1.

Actually by that logic you don't need seconds_per_tick at all

Pickle-Coding replied:
Doesn't seem logical. The idle power usage should be halved instead I guess, so that it's actually truthful.

SyncIt21 commented on code/modules/reagents/chemistry/machinery/chem_dispenser.dm:
```diff
@@ -12,14 +12,12 @@
-	/// Keep track of the intervals made during recharges
-	var/recharge_counter = 0
+	/// The rate at which this machine recharges the power cell.
+	var/recharge_amount = 1.25e3
```
1.25 JOULES.

SyncIt21 commented:
> qol: Chem dispenser UI and crap.

This doesn't explain anything. Write it better

Pickle-Coding commented:
> > qol: Chem dispenser UI and crap.
> 
> 
> 
> This doesn't explain anything. Write it better

I will when I finish the PR.

EEASAS commented:
soul

Pickle-Coding committed:
Proc for charging cell.

Adds a return_power() proc that reduces the apc's load. Mainly used to give back power that wasn't used, such as for charging cells near max charge.

Adds a /obj/machinery proc named charge_cell() to charge a cell. The proc draws from the machine's apc's grid to charge the cell.  Does not draw energy from nothing, and does not waste energy.

Replaced many instances of giving power to the cell with the charge_cell() proc.

The microwave cell charging had weird math. I just replaced it with my own because it appeared to be prone to causing energy duplication issues.

More arbitrary energy unit to joule conversions are done.

Fixes typo I made for the force parameter in the ammo_pack's battery's use proc in the energy minigun's fire process proc. Also makes it always try to fill the cell full, just to prevent any accidental wrong charging rates if their energy cost changes.

Pickle-Coding committed:
A lot of shit.

APCs now only handle consuming static power usage during processing. /obj/machinery/proc/use_power() now handles the consumption of available power, and adding the area channel power usage stats. This should let use_power() be more reliable at conserving energy. It prioritises grid before drawing from cell.

Adds draw_power() proc for APCs. Handles drawing power from the grid and APC cell. area/AddStaticPower() will make the apc call draw_power() so static power consumers will add load to grid or drain APC cell.

Adds back defines for WATT, JOULE, WATTS and JOULES. Mainly just to create a strong definition for the standard energy and power units for the codebase.

Cleans up the power related helper procs for the maths.dm helpers file. energy_to_power() and power_to_energy() converts the units based on the SSmachines wait. display_joules() got renamed to display_energy(). display_power() by default converts the unit from energy to power, as energy is stored in energy and not power when we read it.

Chem dispenser UI now uses prefixed units for the charge display. It also shows energy cost per unit instead of power efficiency for the examine.

Powernets now store energy instead of power. Things that access the power of the powernets will use the energy_to_power() helper.

SMES will use energy when taking or adding to the powernets. Their input and output rates gets scaled by power_to_energy(). Replaces one letter variable names with better ones. Also removes the SMESRATE define. Never understood the point of it.

Energy accumulators now discharge energy based on seconds_per_tick.

For now I removed the converging supply and demand variables for powernets. The power monitors just access the available and load. Their purpose was to make the console look a lot smoother, but it made it feel slow and unresponsive to sudden power changes. I'll look for a better solution in the future to keep the graph responsive without looking super pointy.

I replaced some instances of area use_power() with the apc's terminal use_power() since the area one is just for tracking its stats now.

Probably a lot more small shit I forgot about.

Pickle-Coding committed:
Skill issue.

When you append a plural to a number, usually that's a multiplication.

Pickle-Coding committed:
Fixes static power drawers always consuming APC cell.

Terminal add_load() didn't return anything, so it always used the APC's cell since grid_used was 0 in /obj/machinery/power/apc/draw_power().

Pickle-Coding committed:
Static power gets added earlier.

Moves APC static power draw to its new early_process() proc. All of the apcs' early_process() procs get called in the machines subsystem before other machine processing.

APCs cell charging will use its own area usage channel. The power draw from this is displayed near the APC's charging status. Adds and modifies defines for this.

Fixes removeStaticPower not converting to energy. Area use_power() proc has been modified to exclude static power channels instead of only including dynamic ones, to account for apc charge channel. The APC charge channel does not have a static varient, so it could not be included as a dynamic one due to defines relying on shifting to convert from one another.

Fixes powersinks not draining an appropriate amount of energy from the APC cells. Also won't gain heat from cells it couldn't drain from (empty).

Undefines the ACCUMULATOR_CONSTANT_COEFFICIENT at the end of the file since it's a local define.

Portable generators and various other power generators will convert their power generation to energy correctly when adding to the grid.

Removes some commented out code. /obj/machinery/charge_cell() now has a channel parameter to specify the power channel.

SMES will output power before receiving. This allows them to be fully charged while outputting power.

Pickle-Coding committed:
Linter fix attempt.

Attempts to fix the linters by running Prettier on Apc.jsx and ChemDispenser.tsx.

SyncIt21 commented on code/modules/power/apc/apc_main.dm:
```diff
@@ -660,8 +644,8 @@
-	if(cell && cell.charge >= 20)
-		cell.use(20)
+	if(cell && cell.charge >= 2e4)
+		cell.use(2e4)
```
Some questions. Why was this changed from 20 to 2e4? Why has everything been multipled by 1000?.

Is it because the smes supplies energy in those amounts? Has solar panels output been buffed to keep up with these needs?

Also it would be nice if you could create defines like kilo joules, mega joules and giga joules and attach them to these values for readibility

Pickle-Coding replied:
Cells use to have an energy unit that was effectively 1kJ. Now they don't, so using 1 from a cell is 1 joule instead of 1kJ.

Pickle-Coding replied:
I might create the defines later if I feel like it. Somehow I missed that part of your comment. You're right and I do think the watt defines could be useful for power consumption defines, so people know whether it's power or energy at a glance.

Pickle-Coding committed:
APC charging.

APCs will now charge after all machines have been processed. APCs will now always try to charge as long as it is set to charging, it's operating, there is any surplus power on the grid, and it's not fully charged.

This removes the chargecount variable as it's no longer used.  Removes commented out code. APCs handle putting themselves in the SSmachines processing list tasks. SSmachines no longer assigns the APCs by getting all of the machines of type and subtypes of APCs.

Pickle-Coding commented on code/controllers/subsystem/machines.dm:
```diff
@@ -79,19 +82,49 @@ SUBSYSTEM_DEF(machines)
+				apc_late_processing -= apc
+				apc.datum_flags &= ~DF_ISPROCESSING
+			if(MC_TICK_CHECK)
+				return
```
I noticed while looking at the vareditor that the APCs' cells are being used by small amounts while others are charging (enough power to supply machines) when load gets maxed by APC charge. Why is this the case?

Pickle-Coding committed:
Merge branch 'master' into pain-and-cringe

Pickle-Coding committed:
Fixes borgs instantly losing power.

I blindly multiplied the time parameter of the use_power() proc for robot shit. That made them consume a thousand times more power than necessary. Also scales robot ionpulse power usage.

Cyborg status tab charge display is now prefixed joules.

Pickle-Coding commented:
I should try to reproduce #75789 and see if it's fixed.

Pickle-Coding committed:
Robotact UI.

Gives proper units for robotact UI cell charge and energy consumption.

Adds a new typescript format function to format in joules. Cyborg uses variable for lamp power consumption, which uses a define.

Pickle-Coding committed:
Corrects engineering SMES initial charge.

Pretty sure they use to start at 75% max charge, so I'm setting it to that.

Pickle-Coding committed:
Maxes engineering SMES charge.

I don't remember what the old power consumption was, but with 406kW base from Meta and conservation of energy fixed, I saw some APCs get low enough to turn off equipment by the 10 minute mark. Seems a little earlier than I remember so I'm buffing it.

Pickle-Coding committed:
Define for SMES charge loss EMP act.

Seems better this way.

SyncIt21 commented on code/modules/power/smes.dm:
```diff
@@ -424,7 +424,7 @@
 
 /obj/machinery/power/smes/engineering
-	charge = 37.5 * STANDARD_CELL_CHARGE // Engineering starts with some charge for singulo //sorry little one, singulo as engine is gone
+	charge = 50 * STANDARD_CELL_CHARGE // Engineering starts with some charge for singulo //sorry little one, singulo as engine is gone
```
Does charge variable actually work.

Upon initialization the machine calls RefreshParts() and that proc will override this value based on the cells charge so maybe you should actually adjust the charge of each individual cell such that the sum of their values equals this?

Pickle-Coding replied:
It only does that if the machine starts with no charge.

Pickle-Coding committed:
Rename use_power() to use_energy().

Renames use_power() to use_energy(), and replaces all occurences of use_power() with use_energy().

Re-adds use_power(). It is use_energy(), but it automatically scales the energy consumption by the subsystem's time.

This should remove ambiguity because use_power() implies it will consume a set amount of power instead of energy.

Pickle-Coding committed:
I guess CRASH() early returns or something.

Removes unnecessary return since CRASH() stops code execution in the proc.

deathrobotpunch commented:
please tell me that using joules / watts is optional

Pickle-Coding committed:
Rename area power_usage to energy_usage.

It's a more accurate name. Also adds a comment for it.

Pickle-Coding committed:
Resolve merge conflict.

Pickle-Coding committed:
Commit that does nothing.

Github wasn't receiving the merge conflict resolution commit after I pushed it for some reason. I'll see if this fixes it.

Pickle-Coding committed:
Fixes modsuits.

Makes modsuit energy and power usage defines scale with the cell energy defines, so we don't run into this problem again.

Also renames some variables and comments to be less misleading.

Pickle-Coding commented on tgui/packages/tgui/interfaces/MODsuit.tsx:
```diff
@@ -781,11 +781,11 @@ const ModuleSection = (props) => {
-                <Table.Cell textAlign="center">{module.idle_power}</Table.Cell>
+                <Table.Cell textAlign="center">{formatPower(module.idle_power)}</Table.Cell>
                 <Table.Cell textAlign="center">
-                  {module.active_power}
+                  {formatPower(module.active_power)}
                 </Table.Cell>
-                <Table.Cell textAlign="center">{module.use_power}</Table.Cell>
+                <Table.Cell textAlign="center">{formatEnergy(module.use_energy)}</Table.Cell>
```
Right now this looks awful due to not enough space, so it wraps to a newline. I don't really know typescript, so I'm not sure what to do about this.

SyncIt21 replied:
Just increase the width of the whole window till is stops wrapping.

Use tgui-dev.bat in your bin directory to allow for dynamic changes to your UI mid game.

Pickle-Coding replied:
Thanks. I'll try that. I assumed increasing the width wouldn't fix it, because extending the UI with my cursor didn't fix it due to the numbers portion of the UI having a fixed width.

Pickle-Coding replied:
It didn't work, even when I applied extra width to the columns.

SyncIt21 commented:
Please resolve merge conflicts before I look at the mod suit UI again

Pickle-Coding committed:
Merge branch 'master' into pain-and-cringe

Pickle-Coding commented:
Fuck you mobile webeditor replacing tabs with space indentation.

Pickle-Coding committed:
Replace spaces with tabs.

Replaces some spaces with tabs. Resolving merge conflicts with mobile moment.

Pickle-Coding committed:
MODsuit.tsx formatting.

Runs prettier on it to make the linters happy.

Pickle-Coding committed:
Merge branch 'pain-and-cringe' of https://github.com/Pickle-Coding/tgstation into pain-and-cringe

SyncIt21 commented:
So the width of each table cell is predetermined by the number of the cells in that row so you can't change it properties to suit your needs.

However you change the content inside your cell and just the width of the cell from there.
Currently it looks like this
![Screenshot (401)](https://github.com/tgstation/tgstation/assets/110812394/74fa945c-405d-4ef2-9091-22fc047319d6)

Yup all clottered, not one bit nice.

So what you can do is wrap your text inside a div element & adjust the size of that as such
![Screenshot (403)](https://github.com/tgstation/tgstation/assets/110812394/330464cc-07cd-46cb-8646-7e8b0c544ebb)

After this make sure to increase the width of your window & now everything is neat & spaced
![Screenshot (402)](https://github.com/tgstation/tgstation/assets/110812394/fe2277f9-2c44-4cb7-afa0-fafc63e001f1)

Of course you can adjust the width & decide the spacing & all, but make sure to update the window width as well so everything is displayed.

Also merge conflicts. Let's get this merged

Pickle-Coding committed:
Merge branch 'master' of https://github.com/Pickle-Coding/tgstation into pain-and-cringe

Pickle-Coding committed:
Forgot to mention.

In my last merge commit, I renamed some variables in the market_telepad.dm file to make it more clear.

Pickle-Coding committed:
Prevents modsuit ui unit wrapping.

Prevents modsuit ui unit wrapping by increasing the width and adding inline-block styles (I have no idea what I'm talking about) with 60px padding.

Also increases the height a little to reduce the chance of needing to scroll.

Pickle-Coding commented:
Thanks. This has resolved the wrapping.

Pickle-Coding committed:
Adds procs for checking available energy.

Adds the available_energy() procs for machines and APCs. It returns the accessible energy the machine can use from the APC, including the cell and grid.

Pickle-Coding committed:
Update circuits.

Scales the circuits energy usages with the STANDARD_CELL_CHARGE. Also formats the energy usage information for the ui notices.

Renames the power_usage_per_input and other crap to energy crap.

Pickle-Coding committed:
Scale get_electrocute_damage().

Scales get_electrocute_damage()'s energy scale with energy_to_power(), so the damage is the same for the same power regardless of machine SS wait now that the grid stores energy instead of power.

Pickle-Coding committed:
Removes Bruh.png.

I accidentally saved that there and it got tracked.

SyncIt21 commented:
Fixes #82031.

Also it would be appreciated if you could create defines like mega and gigajoules and same for watts.

Looking at numbers like 1e3 and 1e6 is really unpleasant

Pickle-Coding commented:
What if I create KILO, MEGA, etc defines amd just do KILO JOULES or whatever?

SyncIt21 commented:
> What if I create KILO, MEGA, etc defines amd just do KILO JOULES or whatever?

Its your preference. either create defines like KILO_JOULE or just KILO and append the joules define. Whatever you decide i just want to see those semantics used.

> I'm not sure what causes https://github.com/tgstation/tgstation/issues/82031 so I'm not certain if this fixes it. Do you know why https://github.com/tgstation/tgstation/issues/82031 happens?

I think https://github.com/tgstation/tgstation/issues/81167#issuecomment-1915559079 & all discussions below will provide some answers

Pickle-Coding committed:
Adds SI prefix value defines. Fixes some crap.

Adds SI prefix value defines as a way to make numbers easier to read. Applies that shit to a bunch of energy numbers.

Changes update_use_energy to update_use_power. That was an accident from doing a mass change.

Fixes directly_use_energy() only returning TRUE (1) instead of amount for areas that don't require power. Don't think anything relies on that right now, but it's better to return the amount of energy for things that rely on how much can be used.

Also adds a force parameter for directly_use_energy() and use_energy().

Makes some machines actually check if it can use the energy before performing an action. This isn't an exhaustive change though. There's probably loads more offendors.

Pickle-Coding committed:
Fixes compile error.

Forgot to replace power_used with energy_used here.

Pickle-Coding committed:
Merge branch 'master' of https://github.com/Pickle-Coding/tgstation into pain-and-cringe

SyncIt21 commented on .github/guides/STYLE.md:
```diff
@@ -457,7 +457,7 @@ deal_damage(10) // Fine! The proc name makes it obvious `10` is the damage...at
 deal_damage(damage = 10) // Redundant, but not prohibited.
 
-use_power(30) // Fine! `30` is obviously something like watts.
+use_energy(30) // Fine! `30` is obviously something like joules.
```
```suggestion
use_energy(30 JOULES) // Use energy in joules.
```

SyncIt21 commented on code/game/machinery/buttons.dm:
```diff
@@ -228,7 +228,7 @@
 		return
 
-	use_power(5)
+	use_energy(5)
```
```suggestion
	use_energy(5 JOULES)
```

SyncIt21 commented on code/game/machinery/cell_charger.dm:
```diff
@@ -136,10 +136,11 @@
-	charging.give(main_draw)
-	use_power(charge_rate / 100) //use a small bit for the charger itself, but power usage scales up with the part tier
+
+	use_energy(main_draw * 0.01) //use a small bit for the charger itself, but power usage scales up with the part tier
```
```suggestion
	use_energy(main_draw * GLOB.CHARGELEVEL) //use a small bit for the charger itself, but power usage scales up with the part tier
```

Since they have the same value. Or maybe the meaning here is different? idk you decide

Pickle-Coding replied:
That's just extra power usage with heat inefficiency or something.

SyncIt21 commented on code/game/machinery/computer/camera.dm:
```diff
@@ -173,7 +173,7 @@
 		last_camera_turf = null
 		playsound(src, 'sound/machines/terminal_off.ogg', 25, FALSE)
-		use_power(0)
+		use_energy(0)
```
Weird!. Not required right?

Pickle-Coding replied:
Didn't even notice that.

SyncIt21 commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -1253,14 +1253,14 @@
 			if(!hasPower() || wires.is_cut(WIRE_OPEN) || (obj_flags & EMAGGED))
 				return FALSE
-			use_power(50)
+			use_energy(50)
```
```suggestion
			use_energy(50 JOULES)
```

SyncIt21 commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -1253,14 +1253,14 @@
 			if(obj_flags & EMAGGED)
 				return FALSE
-			use_power(50)
+			use_energy(50)
```
```suggestion
			use_energy(50 JOULES)
```

SyncIt21 commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -1335,7 +1335,7 @@
 			if(obj_flags & EMAGGED)
 				return FALSE
-			use_power(50)
+			use_energy(50)
```
```suggestion
			use_energy(50 JOULES)
```

SyncIt21 commented on code/game/objects/items/devices/powersink.dm:
```diff
@@ -167,15 +167,12 @@
+		if(istype(terminal.master, /obj/machinery/power/apc))
+			var/obj/machinery/power/apc/apc = terminal.master
+			if(apc.operating && apc.cell)
+				drained = apc.cell.use(50000, force = TRUE)
```
```suggestion
				drained = apc.cell.use(50 KILO JOULES, force = TRUE)
```

SyncIt21 commented on code/game/objects/items/stacks/golem_food/golem_status_effects.dm:
```diff
@@ -168,7 +168,7 @@
 	alert_desc = "You are protected from high pressure and can convert heat damage into power."
 	/// What do we multiply our damage by to convert it into power?
-	var/power_multiplier = 5
+	var/power_multiplier = 5000
```
```suggestion
	var/power_multiplier = 5 KILO WATTS
```

Pickle-Coding replied:
It's a multiplier so we don't need units here.

Pickle-Coding replied:
Should be joules.

SyncIt21 commented on code/modules/cargo/markets/market_telepad.dm:
```diff
@@ -19,10 +19,10 @@
 	var/power_efficiency = 1
 	/// Power used per teleported which gets divided by power_efficiency.
-	var/power_usage_per_teleport = 10000
+	var/energy_usage_per_teleport = 10000
```
```suggestion
	var/energy_usage_per_teleport = 10 KILO JOULES
```

SyncIt21 commented on code/modules/mob/living/basic/festivus_pole.dm:
```diff
@@ -38,7 +38,7 @@
 
 	///how much charge we give off to cells around us when rubbed
-	var/recharge_value = 75
+	var/recharge_value = 75000
```
Watts or joules?

Pickle-Coding replied:
It's an action, so joules.

SyncIt21 commented on code/modules/mob/living/silicon/ai/life.dm:
```diff
@@ -31,7 +31,7 @@
 		var/area/home = get_area(src)
 		if(home.powered(AREA_USAGE_EQUIP))
-			home.use_power(500 * seconds_per_tick, AREA_USAGE_EQUIP)
+			home.apc?.terminal?.use_energy(500 * seconds_per_tick, channel = AREA_USAGE_EQUIP)
```
```suggestion
			home.apc?.terminal?.use_energy((0.5 KILO JOULES) * seconds_per_tick, channel = AREA_USAGE_EQUIP)
```

Pickle-Coding replied:
Because we multiply this by time, it should be WATTS. Also it could just be 500 instead of 0.5 KILO.

Pickle-Coding replied:
Also it's safe to not use parenthesis here.

SyncIt21 commented on code/modules/mob/living/silicon/robot/life.dm:
```diff
@@ -14,14 +14,14 @@
+/mob/living/silicon/robot/proc/use_energy(seconds_per_tick, times_fired)
 	if(cell?.charge)
-		if(cell.charge <= 100)
+		if(cell.charge <= 10000)
```
```suggestion
		if(cell.charge <= (10 KILO JOULES))
```

SyncIt21 commented on code/modules/mob/living/silicon/robot/robot.dm:
```diff
@@ -226,11 +226,9 @@
 		return
 
-	if(cell.charge <= 10)
+	if(!cell.use(1e4))
```
```suggestion
	if(!cell.use(10 KILO JOULES))
```

SyncIt21 commented on code/game/machinery/autolathe.dm:
```diff
@@ -96,7 +96,7 @@
 
 	//we use initial(active_power_usage) because higher tier parts will have higher active usage but we have no benifit from it
-	if(directly_use_power(ROUND_UP((amount_inserted / (MAX_STACK_SIZE * SHEET_MATERIAL_AMOUNT)) * 0.02 * initial(active_power_usage))))
+	if(directly_use_energy(ROUND_UP((amount_inserted / (MAX_STACK_SIZE * SHEET_MATERIAL_AMOUNT)) * 0.02 * initial(active_power_usage))))
```
```suggestion
	if(directly_use_energy(ROUND_UP((amount_inserted / (MAX_STACK_SIZE * SHEET_MATERIAL_AMOUNT)) * 0.02 * initial(active_power_usage) KILO JOULES)))
```

Because this proc uses the apc cell directly & since all cells standard capacity has been scaled up 1e3(or 1 kilo joules) we need to compensate for that

Pickle-Coding replied:
I'll just scale the active_power_usage instead, since we don't need to double apply units. On the topic of active_power_usage, I've seen it both for time based processing and action based. I guess I could make a comment saying the unit is ambiguous.

SyncIt21 commented on code/modules/research/machinery/_production.dm:
```diff
@@ -144,7 +144,7 @@
 
 	//we use initial(active_power_usage) because higher tier parts will have higher active usage but we have no benifit from it
-	if(directly_use_power(ROUND_UP((amount_inserted / (MAX_STACK_SIZE * SHEET_MATERIAL_AMOUNT)) * 0.02 * initial(active_power_usage))))
+	if(directly_use_energy(ROUND_UP((amount_inserted / (MAX_STACK_SIZE * SHEET_MATERIAL_AMOUNT)) * 0.02 * initial(active_power_usage))))
```
```suggestion
	if(directly_use_energy(ROUND_UP((amount_inserted / (MAX_STACK_SIZE * SHEET_MATERIAL_AMOUNT)) * 0.02 * initial(active_power_usage) KILO JOULES)))
```

Same reason as autolathe

Pickle-Coding replied:
Same reason as my comment on the autolathe.

SyncIt21 commented on code/modules/research/machinery/_production.dm:
```diff
@@ -289,7 +289,7 @@
 
 			//we use initial(active_power_usage) because higher tier parts will have higher active usage but we have no benifit from it
-			if(!directly_use_power(ROUND_UP((amount / MAX_STACK_SIZE) * 0.02 * initial(active_power_usage))))
+			if(!directly_use_energy(ROUND_UP((amount / MAX_STACK_SIZE) * 0.02 * initial(active_power_usage))))
```
```suggestion
			if(!directly_use_energy(ROUND_UP((amount / MAX_STACK_SIZE) * 0.02 * initial(active_power_usage) KILO JOULES)))
```

SyncIt21 commented on code/modules/research/machinery/_production.dm:
```diff
@@ -375,7 +375,7 @@
 		return
 
-	if(!is_operational || !directly_use_power(charge_per_item))
+	if(!is_operational || !directly_use_energy(charge_per_item))
```
Make sure `charge_per_item` is in koli joules

SyncIt21 commented on code/modules/research/experimentor.dm:
```diff
@@ -505,7 +505,7 @@
 		if(globalMalf > 76 && globalMalf < 98)
 			visible_message(span_warning("[src] begins to smoke and hiss, shaking violently!"))
-			use_power(500000)
+			use_energy(500000)
```
```suggestion
			use_energy(0.5 MEGA JOULES)
```

Pickle-Coding replied:
I'd do 500 KILO JOULES.

SyncIt21 commented on code/modules/research/experimentor.dm:
```diff
@@ -235,7 +235,7 @@
 
 	experiment(reaction, loaded_item)
-	use_power(750)
+	use_energy(750)
```
```suggestion
	use_energy(7.5 KILO JOULES)
```

Pickle-Coding replied:
This is just 750 JOULES.

SyncIt21 commented on code/modules/research/xenobiology/xenobio_camera.dm:
```diff
@@ -212,7 +212,7 @@ Due to keyboard shortcuts, the second one is not necessarily the remote eye's lo
 
 	target_mob.visible_message(span_notice("[target_mob] vanishes as [p_theyre()] reclaimed for recycling!"))
-	connected_recycler.use_power(500)
+	connected_recycler.use_energy(500)
```
```suggestion
	connected_recycler.use_energy(0.5 KILO JOULES)
```

Pickle-Coding replied:
I'll do 500 JOULES when I get on my PC. Same for all other occurences of this.

SyncIt21 commented on code/modules/transport/tram/tram_doors.dm:
```diff
@@ -90,7 +90,7 @@
 		do_sparks(3, TRUE, src)
 		playsound(src, SFX_SPARKS, vol = 75, vary = FALSE, extrarange = SHORT_RANGE_SOUND_EXTRARANGE)
-	use_power(50)
+	use_energy(50)
```
```suggestion
	use_energy(50 JOULES)
```

SyncIt21 commented on code/modules/vehicles/mecha/combat/durand.dm:
```diff
@@ -50,7 +50,7 @@
 /obj/vehicle/sealed/mecha/durand/process()
 	. = ..()
-	if(defense_mode && !use_power(100)) //Defence mode can only be on with a occupant so we check if one of them can toggle it and toggle
+	if(defense_mode && !use_energy(100)) //Defence mode can only be on with a occupant so we check if one of them can toggle it and toggle
```
```suggestion
	if(defense_mode && !use_energy(0.1 KILO JOULES)) //Defence mode can only be on with a occupant so we check if one of them can toggle it and toggle
```

Pickle-Coding replied:
I'll also scale this since this uses the cell.

SyncIt21 reviewed (changes requested):
Some nit picks. Also now since you have buffed up the standard capacity of a cell from 1000->1000000(1 Mega Joules), You need to correctly update all uses of `directly_use_energy()` by scaling with 1000(1 KILO JOULES) So they use the same energy as before

Pickle-Coding committed:
Update STYLE.md

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update buttons.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update airlock.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update airlock.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update airlock.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update powersink.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update market_telepad.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update life.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update robot.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Update tram_doors.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding commented:
Forgot about that. Thanks for the help.

Pickle-Coding committed:
Fix lathe energy consumption + reviews.

Scales lathe energy consumption to match with cells. Also makes active_power_consumption more clear for these machines by making it the energy cost of a full stack of sheets worth of materials spent.

Pickle-Coding committed:
Merge branch 'pain-and-cringe' of https://github.com/Pickle-Coding/tgstation into pain-and-cringe

SyncIt21 commented on code/modules/research/machinery/_production.dm:
```diff
@@ -1,6 +1,8 @@
 	name = "technology fabricator"
 	desc = "Makes researched and prototype items with materials and energy."
+	// Energy cost per full stack of materials spent. Material insertion is 40% of this.
+	active_power_usage = 20 KILO * BASE_MACHINE_ACTIVE_CONSUMPTION
```
Isn't this value a bit too high? For autolathe you did just 25 * active consumption with the same comment so maybe that should apply here as well?

You can copy it's exact code for power consumption and make them match

Pickle-Coding replied:
That was a brain fart. I multiplied by 20 instead of divide.

Pickle-Coding commented on code/modules/research/machinery/_production.dm:
```diff
@@ -1,6 +1,8 @@
 	name = "technology fabricator"
 	desc = "Makes researched and prototype items with materials and energy."
+	// Energy cost per full stack of materials spent. Material insertion is 40% of this.
+	active_power_usage = 20 KILO * BASE_MACHINE_ACTIVE_CONSUMPTION
```
```suggestion
	active_power_usage = 50 * BASE_MACHINE_ACTIVE_CONSUMPTION
```

Pickle-Coding committed:
Update _production.dm

Pickle-Coding committed:
Merge branch 'master' into pain-and-cringe

SyncIt21 commented on code/modules/modular_computers/computers/item/computer.dm:
```diff
@@ -456,7 +456,7 @@
 	if(internal_cell == gone)
 		internal_cell = null
-		if(enabled && !use_power())
+		if(enabled && !use_energy())
```
`use_energy()` with no arguments? This should take some positive value right? Same question for everywhere else in this file

Pickle-Coding replied:
I guess they thought use_power() meant something else. What a strange use case. I'll add arguments or remove these once I find out what's going on in there.

Pickle-Coding replied:
Seems like they have their own proc. Not really relevant to ours and it just got caught up in the mass proc name change. I guess it doesn't matter, though it's confusing. Maybe for a separate PR.

SyncIt21 commented on code/modules/modular_computers/computers/machinery/modular_computer.dm:
```diff
@@ -129,7 +129,7 @@
 // Modular computers can have battery in them, we handle power in previous proc, so prevent this from messing it up for us.
 /obj/machinery/modular_computer/power_change()
-	if(cpu?.use_power()) // If it still has a power source, PC wouldn't go offline.
+	if(cpu?.use_energy()) // If it still has a power source, PC wouldn't go offline.
```
Interesting? no positive value supplied to `use_energy()` & yet it has always worked?

Pickle-Coding replied:
Same as above.

SyncIt21 commented on code/modules/power/lighting/light.dm:
```diff
@@ -472,7 +472,7 @@
 		return FALSE
 	var/obj/item/stock_parts/cell/real_cell = get_cell()
-	if(real_cell.charge > 300) //it's meant to handle 120 W, ya doofus
+	if(real_cell.charge > 3e5) //it's meant to handle 120 W, ya doofus
```
```suggestion
	if(real_cell.charge > 0.3 MEGA JOULES)
```

Pickle-Coding replied:
I made it be based on the initial max charge of the emergency light fixture cell class.

SyncIt21 commented on code/modules/power/power.dm:
```diff
@@ -132,28 +132,100 @@
+ * - ignore_apc: If true, do not consider the APC's cell when demanding power.
+ * Returns: The amount of energy used.
+ */
+/obj/machinery/proc/use_power(amount, channel = power_channel, ignore_apc = FALSE, force = TRUE)
```
I haven't seen this proc used anywhere till now. Did i miss a use case or when do we need this? because in your PR body you did specify everything draws power from the grid as energy.

Pickle-Coding replied:
I guess we don't need this. I just left that there handy in case I got in an odd scenario where I wanted power consumption in some process without access to seconds_per_tick.

SyncIt21 replied:
Well if you have no need for this proc in your PR i suggest you remove it. Writing helper procs for potential future use cases is how dead code is born

Pickle-Coding replied:
Removed.

SyncIt21 commented on code/modules/power/powernet.dm:
```diff
@@ -84,18 +82,14 @@
-	if(avail >= 1000)
-		return clamp(20 + round(avail/25000), 20, 195) + rand(-5,5)
+	if(energy_to_power(avail) >= 1000)
+		return clamp(20 + round(energy_to_power(avail)/25000), 20, 195) + rand(-5,5)
```
What exactly is 25000? it looks like watts but maybe it should have its own define or at least a comment explaining what this magic number is.

Pickle-Coding replied:
Watts per damage, so WATTS would be the closest we have. I'll turn it into a define when I have the time.

Pickle-Coding replied:
Actually it would make more sense if it was joules, and we just assume 1 second of contact, since this exact same code is used somewhere else for cells. I could turn this into a macro.

Pickle-Coding replied:
Added defines and a macro for it.

SyncIt21 commented on code/modules/power/tesla/coil.dm:
```diff
@@ -90,8 +87,8 @@
 	. = ..()
-	zap_sound_volume = min(energy_to_joules(stored_energy)/200000, 100)
-	zap_sound_range = min(energy_to_joules(stored_energy)/4000000, 10)
+	zap_sound_volume = min(processed_energy / 8000, 100)
+	zap_sound_range = min(processed_energy / 1.6e5, 10)
```
A comment explaining what `8000` & `1.6e5` are used here would be helpful

Pickle-Coding replied:
Watts per volume and watts per sound range.

Pickle-Coding replied:
I was wrong. It's actually joules.

Pickle-Coding replied:
I made it power based and used prefixed units and comments.

SyncIt21 reviewed (commented):
Almost there. Some anomalies & questions mostly

Pickle-Coding committed:
Reviews, macros and shit.

Adds a define for energy per damage. Adds a macro for calculating electrocution damage from energy. Uses that instead of a repeated calculation for grid and cell electrocution damage calculation.

Renames some instances of use_power() with use_energy() (pull skew). Also makes those instances check if it can supply the energy demand before doing whatever they do.

Removes use_power() proc. It was unused, and probably never will be given the odd scenario that would call for it.

Light emergency power usage is based on the /obj/item/stock_parts/cell/emergency_light maxcharge variable instead of a magic number.

Tesla coils sound volume and range is based on power instead of energy, so it doesn't change based on machine subsystem process rate.

Pickle-Coding committed:
Merge branch 'pain-and-cringe' of https://github.com/Pickle-Coding/tgstation into pain-and-cringe

SyncIt21 commented on code/modules/mod/modules/modules_antag.dm:
```diff
@@ -160,7 +160,7 @@
 	icon_state = "battlemage_shield"
 	idle_power_cost = DEFAULT_CHARGE_DRAIN * 0 //magic
-	use_power_cost = DEFAULT_CHARGE_DRAIN * 0 //magic too
+	use_energy_cost = DEFAULT_CHARGE_DRAIN * 0 //magic too
```
Whaa? this is 0 right?

SyncIt21 commented on code/modules/power/smes.dm:
```diff
@@ -145,45 +144,45 @@
-			var/obj/structure/cable/N = T.get_cable_node(terminal_cable_layer) //get the connecting node cable, if there's one
-			if (prob(50) && electrocute_mob(usr, N, N, 1, TRUE)) //animate the electrocution if uncautious and unlucky
+			var/obj/structure/cable/connected_cable = turf.get_cable_node(terminal_cable_layer) //get the connecting node cable, if there's one
+			if (prob(50) && electrocute_mob(usr, connected_cable, connected_cable, 1, TRUE)) //animate the electrocution if uncautious and unlucky
```
```suggestion
			if (prob(50) && electrocute_mob(user, connected_cable, connected_cable, 1, TRUE)) //animate the electrocution if uncautious and unlucky
```

user provided by `attackby`

SyncIt21 reviewed (commented):
Looks good. This should be my last reviews.

Make sure to address your own comment in https://github.com/tgstation/tgstation/pull/81579#discussion_r1501353312

Pickle-Coding committed:
Update smes.dm

Co-authored-by: SyncIt21 <110812394+SyncIt21@users.noreply.github.com>

Pickle-Coding committed:
Silly shit.

Multiplying by 0 is just 0, so lets just type 0.

Pickle-Coding committed:
Merge branch 'pain-and-cringe' of https://github.com/Pickle-Coding/tgstation into pain-and-cringe

Pickle-Coding commented:
Seems like it's modular computers that is causing the APC power drain, though I'm not sure why that's the cause at the moment.

Pickle-Coding committed:
Whoops!

Should add rather than set lol.

Pickle-Coding committed:
Whoops 2.

It's only the heat from the apc drain that needs to get scaled to match with previous  behaviour.

Pickle-Coding commented:
Found a rare but potentially catastrophic bug relating to the controller, so I am drafting this for now until I find a solution.

Pickle-Coding committed:
I am mega silly.

I was assigning the currentrun reference to a different list instead of assigning currentrun to that list. This caused strange phenomena when an early return from MC_TICK_CHECK occurs.

Pickle-Coding commented:
> Found a rare but potentially catastrophic bug relating to the controller, so I am drafting this for now until I find a solution.

Resolved.

Ghommie commented on code/__DEFINES/si.dm:
```diff
@@ -0,0 +1,21 @@
+#define KILO * 1e3
+#define MEGA * 1e6
+#define GIGA * 1e9
+#define TERRA * 1e12
```
```suggestion
#define TERA * 1e12
```

Ghommie commented on code/modules/power/apc/apc_main.dm:
```diff
@@ -730,6 +689,23 @@
+
+/// Draws energy from the connected grid. When there isn't enough surplus energy from the grid, draws the rest of the demand from its cell. Returns the energy used.
+/obj/machinery/power/apc/proc/draw_energy(amount)
+	var/grid_used = min(terminal.surplus(), amount)
+	terminal?.add_load(grid_used)
```
```suggestion
	var/grid_used = min(terminal?.surplus(), amount)
	terminal?.add_load(grid_used)
```

Ghommie commented on code/modules/power/power.dm:
```diff
@@ -132,28 +132,85 @@
+/obj/machinery/proc/available_energy(consider_cell = TRUE)
+	var/area/home = get_area(src)
+
+	if(!home)
```
```suggestion
	if(isnull(home))
```

Pickle-Coding replied:
I'll commit it, but why should I explicitly null check this?

Ghommie commented on code/modules/power/power.dm:
```diff
@@ -132,28 +132,85 @@
+		return FALSE
+	var/area/home = get_area(src)
+
+	if(!home)
```
```suggestion
	if(isnull(home))
```

Ghommie commented on code/modules/power/power.dm:
```diff
@@ -132,28 +132,85 @@
+		return amount //Shuttles get free power, don't ask why
+
+	var/obj/machinery/power/apc/local_apc = home.apc
+	if(!local_apc)
```
```suggestion
	if(isnull(local_apc))
```

Ghommie commented on code/modules/power/power.dm:
```diff
@@ -132,28 +132,85 @@
+		return INFINITY
+
+	var/obj/machinery/power/apc/local_apc = home.apc
+	if(!local_apc)
```
```suggestion
	if(isnull(local_apc))
```

Pickle-Coding replied:
Same for APCs. Why is it better to explicitly null check them?

Ghommie commented on code/modules/power/power.dm:
```diff
@@ -469,3 +540,5 @@
 			C.update_appearance() // I hate this. it's here because update_icon_state SCANS nearby turfs for objects to connect to. Wastes cpu time
 			return C
 	return null
+
+
```
```suggestion
```

Pickle-Coding committed:
Update si.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

Pickle-Coding committed:
Update apc_main.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

Pickle-Coding committed:
Update power.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

Pickle-Coding committed:
Update power.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

Pickle-Coding committed:
Update power.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

Pickle-Coding committed:
Update power.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

Pickle-Coding committed:
Update power.dm

Co-authored-by: Ghom <42542238+Ghommie@users.noreply.github.com>

SyncIt21 commented:
Amen. Well done @Pickle-Coding. Hopefully this won't cause a ton of new issues to open up but even if it does, I'll be there to help out 👍

Pickle-Coding commented:
Only issue I've seen so far, though it might not be related to this PR, is power consoles and cable multitool/ghost examine not always showing an accurate load. I think this might be related to machines SS yielding after the tick check, then examining before every machine processes. Maybe I could cache the last load power supply and load and return that instead for powernet examines.
